src/Model/User.php

```diff
@@ -78,21 +78,105 @@ public function __construct( \WP_User $user ) {
+		 *
+		 * Otherwise we need to continue and check if the User should be considered private or not
+		 */
+		if ( null !== $is_private ) {
+			return (bool) $is_private;
+		}
+
+		/**
+		 * If the requesting user does not have "list_users" capabilities and the current user is not
+		 * the user being requested, we need to check if the user being requested has published posts
+		 * which makes them a non-private entity.
+		 */
 		if ( ! current_user_can( 'list_users' ) && false === $this->owner_matches_current_user() ) {
 
 			/**
-			 * @todo: We should handle this check in a Deferred resolver. Right now it queries once per user
-			 *      but we _could_ query once for _all_ users.
-			 *
-			 *      For now, we only query if the current user doesn't have list_users, instead of querying
-			 *      for ALL users. Slightly more efficient for authenticated users at least.
+			 * Get allowed Post Types
+			 */
+			$post_types = \WPGraphQL::get_allowed_post_types();
+			unset( $post_types['revision'] );
+			unset( $post_types['attachment'] );
+
+			/**
+			 * If the user has no published posts of any allowed post type,
+			 * they are considered to be private entities.
 			 */
-			if ( ! count_user_posts( absint( $this->data->ID ), \WPGraphQL::get_allowed_post_types(), true ) ) {
-				return true;
+			if ( 0 === $this->count_user_posts( absint( $this->data->ID ), $post_types, true ) ) {
+				$is_private = true;
 			}
 		}
 
-		return false;
+		return $is_private;
+
+	}
+
+	/**
+	 * Cached version (and variation) of count_user_posts, which is uncached but doesn't always
+	 * need to hit the db
+	 *
+	 * count_user_posts is generally fast on smaller sites, but slows on large data sets.
+	 *
+	 * It can also be easy to end up with many redundant queries if it's called several times per
+	 * request. This allows bypassing the db queries in favor of the cache
+	 *
+	 * @param int @userid ID of the User
+	 * @param mixed|string|array         $post_type   The post type(s) to check for published posts in
+	 * @param boolean                    $public_only Whether to count only publicly published posts
+	 *
+	 * @return int|mixed|boolean
+	 */
+	protected function count_user_posts( $userid, $post_type = 'post', $public_only = false ) {
+
+		if ( ! is_numeric( $userid ) ) {
+			return 0;
+		}
+
+		$cache_key   = 'graphql_user_' . (int) $userid;
+		$cache_group = 'user_posts_count';
+
+		/**
+		 * If there's no cache, query to see if the author has 1 published post.
+		 *
+		 * This method of querying is more efficient on large datasets
+		 * then the core count_user_posts() method as this stops at the first match,
+		 * where the count_user_psts() method needs to find all matches and
+		 * it can take a while to calculate.
+		 *
+		 * We only care if there's at least 1 published post, not the total.
+		 */
+		if ( false === ( $count = wp_cache_get( $cache_key, $cache_group ) ) ) {
```

Comment:
This falls apart if someone un-publishes the post. We would have to have some code hooked into transition post status to bust this cache.

Also, where does this cache get set? I only see the get call.

Reply:
ya. . .that's what I was thinking too. 😬 

Where's DFM Transients when you need them?

Reply:
> Also, where does this cache get set? I only see the get call.

uhhh. . .looks like I caught a case of the "moving too fasts" 🤢

Reply:
You will also get stuck in a bad state if an author has 0 published posts, and then publishes a post, as it will still be cached as `0`. So you'll need to bust the cache in that instance as well.

We definitely want to cache `0` as a legitimate result since querying for published posts for a user that has none will do a full table scan to result in the `0`.

Reply:
i would consider calling to `wpcom_vip_count_user_posts` behind a `function_exists` check. that function basically implements what you've done here but i believe they handle the cache busting:

https://github.com/svn2github/wordpress-vip-plugins/blob/master/vip-do-not-include-on-wpcom/wpcom-caching.php#L410

i don't have a strong opinion on needing to bust the cache, i'd probably lean towards not busting it and just documenting the delay.

can i ask what the threat model is here? why exactly do users without published posts have an expectation of privacy? the whole reason that this method is unperformant is that it _always_ runs, even when the user connected to a post (in which case, they obviously have pubbed a post).

is there a way to make this run conditional depending on the `$source` object?